presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java

```diff
@@ -129,42 +135,43 @@ private static InternalAggregationFunction generateAggregation(Type type)
         return ImmutableList.of(new ParameterMetadata(STATE), new ParameterMetadata(BLOCK_INPUT_CHANNEL, type), new ParameterMetadata(BLOCK_INDEX));
     }
 
-    public static void inputShortDecimal(Type type, BigIntegerAndLongState state, Block block, int position)
+    public static void inputShortDecimal(Type type, LongDecimalWithOverflowAndLongState state, Block block, int position)
     {
-        accumulateValueInState(BigInteger.valueOf(type.getLong(block, position)), state);
+        accumulateValueInState(UnscaledDecimal128Arithmetic.unscaledDecimal(type.getLong(block, position)), state);
     }
 
-    public static void inputLongDecimal(Type type, BigIntegerAndLongState state, Block block, int position)
+    public static void inputLongDecimal(Type type, LongDecimalWithOverflowAndLongState state, Block block, int position)
     {
-        accumulateValueInState(decodeUnscaledValue(type.getSlice(block, position)), state);
+        accumulateValueInState(type.getSlice(block, position), state);
     }
 
-    private static void accumulateValueInState(BigInteger value, BigIntegerAndLongState state)
+    private static void accumulateValueInState(Slice unscaledValue, LongDecimalWithOverflowAndLongState state)
     {
-        initializeIfNeeded(state);
-        state.setBigInteger(state.getBigInteger().add(value));
+        accumulateAndUpdateOverflow(unscaledValue, state);
         state.setLong(state.getLong() + 1);
     }
 
-    private static void initializeIfNeeded(BigIntegerAndLongState state)
+    private static void initializeIfNeeded(LongDecimalWithOverflowAndLongState state)
     {
-        if (state.getBigInteger() == null) {
-            state.setBigInteger(BigInteger.valueOf(0));
+        if (state.getLongDecimal() == null) {
+            state.setLongDecimal(UnscaledDecimal128Arithmetic.unscaledDecimal());
         }
     }
 
-    public static void combine(BigIntegerAndLongState state, BigIntegerAndLongState otherState)
+    public static void combine(LongDecimalWithOverflowAndLongState state, LongDecimalWithOverflowAndLongState otherState)
     {
         state.setLong(state.getLong() + otherState.getLong());
-        if (state.getBigInteger() == null) {
-            state.setBigInteger(otherState.getBigInteger());
+        state.setOverflow(state.getOverflow() + otherState.getOverflow());
```

Comment:
Why not? They do and that's the whole point of this overflow: that you can aggregate average out of 10000000 MAX_VALUEs. Those tests are passing:
```
    @Test
    public void testGrzybek1()
    {
        addToState(state, Decimals.MAX_DECIMAL_UNSCALED_VALUE);
        addToState(state, Decimals.MAX_DECIMAL_UNSCALED_VALUE);

        LongDecimalWithOverflowState otherState = new LongDecimalWithOverflowStateFactory().createSingleState();

        addToState(otherState, Decimals.MIN_DECIMAL_UNSCALED_VALUE);
        addToState(otherState, Decimals.MIN_DECIMAL_UNSCALED_VALUE);
        addToState(otherState, Decimals.MIN_DECIMAL_UNSCALED_VALUE);

        DecimalSumAggregation.combine(state, otherState);
        assertEquals(state.getOverflow(), 0);
        assertEquals(state.getLongDecimal(), unscaledDecimal(Decimals.MIN_DECIMAL_UNSCALED_VALUE));
    }
```
```
    @Test
    public void testGrzybek2()
    {
        addToState(state, MIN_DECIMAL_UNSCALED_VALUE);
        addToState(state, MIN_DECIMAL_UNSCALED_VALUE.negate());

        LongDecimalWithOverflowState otherState = new LongDecimalWithOverflowStateFactory().createSingleState();

        addToState(otherState, MAX_DECIMAL_UNSCALED_VALUE);
        addToState(otherState, MAX_DECIMAL_UNSCALED_VALUE);
        addToState(otherState, MAX_DECIMAL_UNSCALED_VALUE);

        DecimalSumAggregation.combine(state, otherState);
        assertEquals(state.getOverflow(), 1);
        assertEquals(
                UnscaledDecimal128Arithmetic.toStringDestructive(state.getLongDecimal()),
                MIN_DECIMAL_UNSCALED_VALUE.add(MIN_DECIMAL_UNSCALED_VALUE.negate())
                        .add(MAX_DECIMAL_UNSCALED_VALUE).add(MAX_DECIMAL_UNSCALED_VALUE).add(MAX_DECIMAL_UNSCALED_VALUE)
                        .subtract(TWO.pow(127)).toString());
    }
```

In this second test, value `TWO.pow(127)` represents `overflow = 1`.